### DISCUSSION TITLE
set functions in rendered elements by mapped over state not always working as intended

### DISCUSSION THREAD

**Discussion by WE-Kaito** [neuefische/questions#581] [Web] [answered]
In my [calorie & workout tracker](https://github.com/WE-dev1/calorie-and-workout-tracker-neuefische-capstone) I use an exercises state [{id = uid() , workout, title, sets, reps, weight time},...]. On the exercises page(page/workouts/[index].js) and in the todays workout moodle(/components/Calendar/CalendarExercises.js) the exercises array is logged in the console. When switching again to a workout and trying to set or delete exercises, ids are randomly copied to other exercise objects. I couldn't find the cause of the issue, it is working as intended when trying the functionality with a clean storage.

**Comment by ThomasF85** [comment, marked as answer]
Hi Enrico,

i cannot reproduce the error you are describing, but I get a Typeerror when trying to delete a Workout.

You have a major problem in your App: You are identifying workouts by their index in an array. Your page `page/workouts/[index].js` uses an index as identifier where it should use an id.
The index changes every time you add or delete an exercise, it is not stable and not unique. The error I get and probably your other errors too arise from that problem.

Another problem is that you hand over an array of filtered Exercises to all your pages. This makes your code hard to understand and debugging more complicated. Filtering also changes the index of items in the filtered array. Here is the code in your `_app.js`:
```
export default function App({ Component, pageProps }) {
  const { exercises } = useCalorieStore();
  const filteredExercises = exercises
    .slice()
    .filter(
      (exercise, index, self) =>
        index === self.findIndex((e) => e.workout === exercise.workout)
    );

  return (
    <>
      <Head>
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0"
        />
      </Head>
      <GlobalStyle />
      <Component {...pageProps} workouts={filteredExercises} />
    </>
  );
}
```
Only your workout pages need the workout array that you hand over. Let your workout pages get the workout from the store directly and not pass a filtered version of them as props in your `_app.js`.

In your workouts page you also check for the existence of an element with index 0 and then assume the existence of another item:
```
  if (workouts[0])
    return (
      <Wrapper>
        <h1
          style={{
            position: "absolute",
            top: "85px",
            left: "70px",
            zIndex: "10",
            color: "var(--3)",
            fontSize: 28,
          }}
        >
          {workouts[index].workout}
```
This is what caused the bug I experienced and it will automatically be fixed when using ids as identifiers (that's what they are) instead of indizes.
